<p style="text-align:center">
    <a href="https://skills.network" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>


# **SpaceX  Falcon 9 first stage Landing Prediction**


# Lab 1: Collecting the data


Estimated time needed: **45** minutes


In this capstone, we will predict if the Falcon 9 first stage will land successfully. SpaceX advertises Falcon 9 rocket launches on its website with a cost of 62 million dollars; other providers cost upward of 165 million dollars each, much of the savings is because SpaceX can reuse the first stage. Therefore if we can determine if the first stage will land, we can determine the cost of a launch. This information can be used if an alternate company wants to bid against SpaceX for a rocket launch. In this lab, you will collect and make sure the data is in the correct format from an API. The following is an example of a successful and launch.


![](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/lab_v2/images/landing_1.gif)


Several examples of an unsuccessful landing are shown here:


![](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/lab_v2/images/crash.gif)


Most unsuccessful landings are planned. Space X performs a controlled landing in the oceans. 


## Objectives


In this lab, you will make a get request to the SpaceX API. You will also do some basic data wrangling and formating. 

- Request to the SpaceX API
- Clean the requested data


----


## Import Libraries and Define Auxiliary Functions


We will import the following libraries into the lab


In [1]:
# Requests allows us to make HTTP requests which we will use to get data from an API
import requests
# Pandas is a software library written for the Python programming language for data manipulation and analysis.
import pandas as pd
# NumPy is a library for the Python programming language, adding support for large, multi-dimensional arrays and matrices, along with a large collection of high-level mathematical functions to operate on these arrays
import numpy as np
# Datetime is a library that allows us to represent dates
import datetime

# Setting this option will print all collumns of a dataframe
pd.set_option('display.max_columns', None)
# Setting this option will print all of the data in a feature
pd.set_option('display.max_colwidth', None)

Below we will define a series of helper functions that will help us use the API to extract information using identification numbers in the launch data.

From the <code>rocket</code> column we would like to learn the booster name.


In [2]:
# Takes the dataset and uses the rocket column to call the API and append the data to the list
def getBoosterVersion(data):
    for x in data['rocket']:
        if x:
            response = requests.get("https://api.spacexdata.com/v4/rockets/"+str(x)).json()
            BoosterVersion.append(response['name'])

From the <code>launchpad</code> we would like to know the name of the launch site being used, the logitude, and the latitude.


In [3]:
# Takes the dataset and uses the launchpad column to call the API and append the data to the list
def getLaunchSite(data):
    for x in data['launchpad']:
        if x:
            response = requests.get("https://api.spacexdata.com/v4/launchpads/"+str(x)).json()
            Longitude.append(response['longitude'])
            Latitude.append(response['latitude'])
            LaunchSite.append(response['name'])

From the <code>payload</code> we would like to learn the mass of the payload and the orbit that it is going to.


In [4]:
# Takes the dataset and uses the payloads column to call the API and append the data to the lists
def getPayloadData(data):
    for load in data['payloads']:
        if load:
            response = requests.get("https://api.spacexdata.com/v4/payloads/"+load).json()
            PayloadMass.append(response['mass_kg'])
            Orbit.append(response['orbit'])

From <code>cores</code> we would like to learn the outcome of the landing, the type of the landing, number of flights with that core, whether gridfins were used, wheter the core is reused, wheter legs were used, the landing pad used, the block of the core which is a number used to seperate version of cores, the number of times this specific core has been reused, and the serial of the core.


In [5]:
# Takes the dataset and uses the cores column to call the API and append the data to the lists
def getCoreData(data):
    for core in data['cores']:
        if core['core'] is not None:
            response = requests.get("https://api.spacexdata.com/v4/cores/"+core['core']).json()
            Block.append(response['block'])
            ReusedCount.append(response['reuse_count'])
            Serial.append(response['serial'])
        else:
            Block.append(None)
            ReusedCount.append(None)
            Serial.append(None)
            Outcome.append(str(core['landing_success'])+' '+str(core['landing_type']))
            Flights.append(core['flight'])
            GridFins.append(core['gridfins'])
            Reused.append(core['reused'])
            Legs.append(core['legs'])
            LandingPad.append(core['landpad'])

Now let's start requesting rocket launch data from SpaceX API with the following URL:


In [6]:
spacex_url="https://api.spacexdata.com/v4/launches/past"

In [7]:
response = requests.get(spacex_url)

Check the content of the response


In [8]:
#print(response.content)

You should see the response contains massive information about SpaceX launches. Next, let's try to discover some more relevant information for this project.


### Task 1: Request and parse the SpaceX launch data using the GET request


To make the requested JSON results more consistent, we will use the following static response object for this project:


In [9]:
static_json_url='https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/API_call_spacex_api.json'

We should see that the request was successfull with the 200 status response code


In [10]:
response.status_code

200

Now we decode the response content as a Json using <code>.json()</code> and turn it into a Pandas dataframe using <code>.json_normalize()</code>


In [11]:
# Use json_normalize meethod to convert the json result into a dataframe
data = pd.json_normalize(response.json())

Using the dataframe <code>data</code> print the first 5 rows


In [12]:
# Get the head of the dataframe
data.head()

static_fire_date_utc  static_fire_date_unix    net  window  \
0  2006-03-17T00:00:00.000Z           1.142554e+09  False     0.0   
1                      None                    NaN  False     0.0   
2                      None                    NaN  False     0.0   
3  2008-09-20T00:00:00.000Z           1.221869e+09  False     0.0   
4                      None                    NaN  False     0.0   

                     rocket success  \
0  5e9d0d95eda69955f709d1eb   False   
1  5e9d0d95eda69955f709d1eb   False   
2  5e9d0d95eda69955f709d1eb   False   
3  5e9d0d95eda69955f709d1eb    True   
4  5e9d0d95eda69955f709d1eb    True   

                                                                                                            failures  \
0                                                [{'time': 33, 'altitude': None, 'reason': 'merlin engine failure'}]   
1            [{'time': 301, 'altitude': 289, 'reason': 'harmonic oscillation leading to premature engine shutdown'}]   
2  [{'time': 140, 'altitude': 35, 'reason': 'residual stage-1 thrust led to collision between stage 1 and stage 2'}]   
3                                                                                                                 []   
4                                                                                                                 []   

                                                                                                                                                                                details  \
0                                                                                                                                      Engine failure at 33 seconds and loss of vehicle   
1  Successful first stage burn and transition to second stage, maximum altitude 289 km, Premature engine shutdown at T+7 min 30 s, Failed to reach orbit, Failed to recover first stage   
2                                                                                                                  Residual stage 1 thrust led to collision between stage 1 and stage 2   
3                        Ratsat was carried to orbit on the first successful orbital launch of any privately funded and developed, liquid-propelled carrier rocket, the SpaceX Falcon 1   
4                                                                                                                                                                                  None   

  crew ships capsules                                              payloads  \
0   []    []       []                            [5eb0e4b5b6c3bb0006eeb1e1]   
1   []    []       []                            [5eb0e4b6b6c3bb0006eeb1e2]   
2   []    []       []  [5eb0e4b6b6c3bb0006eeb1e3, 5eb0e4b6b6c3bb0006eeb1e4]   
3   []    []       []                            [5eb0e4b7b6c3bb0006eeb1e5]   
4   []    []       []                            [5eb0e4b7b6c3bb0006eeb1e6]   

                  launchpad  flight_number         name  \
0  5e9e4502f5090995de566f86              1    FalconSat   
1  5e9e4502f5090995de566f86              2      DemoSat   
2  5e9e4502f5090995de566f86              3  Trailblazer   
3  5e9e4502f5090995de566f86              4       RatSat   
4  5e9e4502f5090995de566f86              5     RazakSat   

                   date_utc   date_unix                 date_local  \
0  2006-03-24T22:30:00.000Z  1143239400  2006-03-25T10:30:00+12:00   
1  2007-03-21T01:10:00.000Z  1174439400  2007-03-21T13:10:00+12:00   
2  2008-08-03T03:34:00.000Z  1217734440  2008-08-03T15:34:00+12:00   
3  2008-09-28T23:15:00.000Z  1222643700  2008-09-28T11:15:00+12:00   
4  2009-07-13T03:35:00.000Z  1247456100  2009-07-13T15:35:00+12:00   

  date_precision  upcoming  \
0           hour     False   
1           hour     False   
2           hour     False   
3           hour     False   
4           hour     False   

                                                                                 

You will notice that a lot of the data are IDs. For example the rocket column has no information about the rocket just an identification number.

We will now use the API again to get information about the launches using the IDs given for each launch. Specifically we will be using columns <code>rocket</code>, <code>payloads</code>, <code>launchpad</code>, and <code>cores</code>.


In [13]:
# Lets take a subset of our dataframe keeping only the features we want and the flight number, and date_utc.
data = data[['rocket', 'payloads', 'launchpad', 'cores', 'flight_number', 'date_utc']]

# We will remove rows with multiple cores because those are falcon rockets with 2 extra rocket boosters and rows that have multiple payloads in a single rocket.
data = data[data['cores'].map(len) == 1]
data = data[data['payloads'].map(len) == 1]

# Since payloads and cores are lists of size 1 we will also extract the single value in the list and replace the feature.
data['cores'] = data['cores'].map(lambda x: x[0])
data['payloads'] = data['payloads'].map(lambda x: x[0])

# We also want to convert the date_utc to a datetime datatype and then extracting the date leaving the time
data['date'] = pd.to_datetime(data['date_utc']).dt.date

# Using the date we will restrict the dates of the launches
data = data[data['date'] <= datetime.date(2020, 11, 13)]

* From the <code>rocket</code> we would like to learn the booster name

* From the <code>payload</code> we would like to learn the mass of the payload and the orbit that it is going to

* From the <code>launchpad</code> we would like to know the name of the launch site being used, the longitude, and the latitude.

* **From <code>cores</code> we would like to learn the outcome of the landing, the type of the landing, number of flights with that core, whether gridfins were used, whether the core is reused, whether legs were used, the landing pad used, the block of the core which is a number used to seperate version of cores, the number of times this specific core has been reused, and the serial of the core.**

The data from these requests will be stored in lists and will be used to create a new dataframe.


In [14]:
#Global variables 
BoosterVersion = []
PayloadMass = []
Orbit = []
LaunchSite = []
Outcome = []
Flights = []
GridFins = []
Reused = []
Legs = []
LandingPad = []
Block = []
ReusedCount = []
Serial = []
Longitude = []
Latitude = []

These functions will apply the outputs globally to the above variables. Let's take a looks at <code>BoosterVersion</code> variable. Before we apply  <code>getBoosterVersion</code> the list is empty:


In [15]:
BoosterVersion

[]

Now, let's apply <code> getBoosterVersion</code> function method to get the booster version


In [16]:
# Call getBoosterVersion
getBoosterVersion(data)

the list has now been update 


In [17]:
BoosterVersion[0:5]

['Falcon 1', 'Falcon 1', 'Falcon 1', 'Falcon 1', 'Falcon 9']

we can apply the rest of the  functions here:


In [18]:
# Call getLaunchSite
getLaunchSite(data)

In [19]:
# Call getPayloadData
getPayloadData(data)

In [20]:
# Call getCoreData
getCoreData(data)

Finally lets construct our dataset using the data we have obtained. We we combine the columns into a dictionary.


In [21]:
launch_dict = \
    {'FlightNumber': list(data['flight_number']),
        'Date': list(data['date']),
        'BoosterVersion': BoosterVersion,
        'PayloadMass': PayloadMass,
        'Orbit': Orbit,
        'LaunchSite': LaunchSite,
        'Outcome': Outcome,
        'Flights': Flights,
        'GridFins': GridFins,
        'Reused': Reused,
        'Legs': Legs,
        'LandingPad': LandingPad,
        'Block': Block,
        'ReusedCount': ReusedCount,
        'Serial': Serial,
        'Longitude': Longitude,
        'Latitude': Latitude}

Then, we need to create a Pandas data frame from the dictionary launch_dict.

In [22]:
# Create a data from launch_dict
pd.DataFrame.from_dict(launch_dict, orient='index')

0                1                2   \
FlightNumber                  1                2                4   
Date                 2006-03-24       2007-03-21       2008-09-28   
BoosterVersion         Falcon 1         Falcon 1         Falcon 1   
PayloadMass                  20             None              165   
Orbit                       LEO              LEO              LEO   
LaunchSite      Kwajalein Atoll  Kwajalein Atoll  Kwajalein Atoll   
Outcome                    None             None             None   
Flights                    None             None             None   
GridFins                   None             None             None   
Reused                     None             None             None   
Legs                       None             None             None   
LandingPad                 None             None             None   
Block                      None             None             None   
ReusedCount                   0                0                0   
Serial                 Merlin1A         Merlin2A         Merlin2C   
Longitude               167.743          167.743          167.743   
Latitude                9.04772          9.04772          9.04772   

                             3             4             5             6   \
FlightNumber                  5             6             8            10   
Date                 2009-07-13    2010-06-04    2012-05-22    2013-03-01   
BoosterVersion         Falcon 1      Falcon 9      Falcon 9      Falcon 9   
PayloadMass                 200          None           525           677   
Orbit                       LEO           LEO           LEO           ISS   
LaunchSite      Kwajalein Atoll  CCSFS SLC 40  CCSFS SLC 40  CCSFS SLC 40   
Outcome                    None          None          None          None   
Flights                    None          None          None          None   
GridFins                   None          None          None          None   
Reused                     None          None          None          None   
Legs                       None          None          None          None   
LandingPad                 None          None          None          None   
Block                      None             1             1             1   
ReusedCount                   0             0             0             0   
Serial                 Merlin3C         B0003         B0005         B0007   
Longitude               167.743      -80.5774      -80.5774      -80.5774   
Latitude                9.04772       28.5619       28.5619       28.5619   

                         7             8             9             10  \
FlightNumber             11            12            13            14   
Date             2013-09-29    2013-12-03    2014-01-06    2014-04-18   
BoosterVersion     Falcon 9      Falcon 9      Falcon 9      Falcon 9   
PayloadMass             500          3170          3325          2296   
Orbit                    PO           GTO           GTO           ISS   
LaunchSite      VAFB SLC 4E  CCSFS SLC 40  CCSFS SLC 40  CCSFS SLC 40   
Outcome                None          None          None          None   
Flights                None          None          None          None   
GridFins               None          None          None          None   
Reused                 None          None          None          None   
Legs                   None          None          None          None   
LandingPad             None          None          None          None   
Block                     1             1             1             1   
ReusedCount               0             0             0             0   
Serial                B1003         B1004         B1005         B1006   
Longitude          -120.611      -80.5774      -80.5774      -80.5774   
Latitude            34.6321       28.5619       28.5619       28.5619   

                          11            12            13            14  \
FlightNumb

Show the summary of the dataframe


In [23]:
# Show the head of the dataframe
pd.DataFrame.from_dict(launch_dict, orient='index').transpose().head()

FlightNumber        Date BoosterVersion PayloadMass Orbit       LaunchSite  \
0            1  2006-03-24       Falcon 1          20   LEO  Kwajalein Atoll   
1            2  2007-03-21       Falcon 1        None   LEO  Kwajalein Atoll   
2            4  2008-09-28       Falcon 1         165   LEO  Kwajalein Atoll   
3            5  2009-07-13       Falcon 1         200   LEO  Kwajalein Atoll   
4            6  2010-06-04       Falcon 9        None   LEO     CCSFS SLC 40   

  Outcome Flights GridFins Reused  Legs LandingPad Block ReusedCount  \
0    None    None     None   None  None       None  None           0   
1    None    None     None   None  None       None  None           0   
2    None    None     None   None  None       None  None           0   
3    None    None     None   None  None       None  None           0   
4    None    None     None   None  None       None     1           0   

     Serial Longitude Latitude  
0  Merlin1A   167.743  9.04772  
1  Merlin2A   167.743  9.04772  
2  Merlin2C   167.743  9.04772  
3  Merlin3C   167.743  9.04772  
4     B0003  -80.5774  28.5619

### Task 2: Filter the dataframe to only include `Falcon 9` launches


Finally we will remove the Falcon 1 launches keeping only the Falcon 9 launches. Filter the data dataframe using the <code>BoosterVersion</code> column to only keep the Falcon 9 launches. Save the filtered data to a new dataframe called <code>data_falcon9</code>.


In [24]:
# Hint data['BoosterVersion']!='Falcon 1'
datatemp = pd.DataFrame.from_dict(launch_dict, orient='index').transpose()
data_falcon9 = datatemp[datatemp['BoosterVersion'] != 'Falcon 1']

Now that we have removed some values we should reset the FlgihtNumber column


In [25]:
data_falcon9.loc[:,'FlightNumber'] = list(range(1, data_falcon9.shape[0]+1))
data_falcon9

/home/jupyterlab/conda/envs/python/lib/python3.7/site-packages/pandas/core/indexing.py:1781: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[item_labels[indexer[info_axis]]] = value


FlightNumber        Date BoosterVersion PayloadMass Orbit    LaunchSite  \
4              1  2010-06-04       Falcon 9        None   LEO  CCSFS SLC 40   
5              2  2012-05-22       Falcon 9         525   LEO  CCSFS SLC 40   
6              3  2013-03-01       Falcon 9         677   ISS  CCSFS SLC 40   
7              4  2013-09-29       Falcon 9         500    PO   VAFB SLC 4E   
8              5  2013-12-03       Falcon 9        3170   GTO  CCSFS SLC 40   
..           ...         ...            ...         ...   ...           ...   
89            86  2020-09-03       Falcon 9       15600  VLEO    KSC LC 39A   
90            87  2020-10-06       Falcon 9       15600  VLEO    KSC LC 39A   
91            88  2020-10-18       Falcon 9       15600  VLEO    KSC LC 39A   
92            89  2020-10-24       Falcon 9       15600  VLEO  CCSFS SLC 40   
93            90  2020-11-05       Falcon 9        3681   MEO  CCSFS SLC 40   

   Outcome Flights GridFins Reused  Legs LandingPad Block ReusedCount Serial  \
4     None    None     None   None  None       None     1           0  B0003   
5     None    None     None   None  None       None     1           0  B0005   
6     None    None     None   None  None       None     1           0  B0007   
7     None    None     None   None  None       None     1           0  B1003   
8     None    None     None   None  None       None     1           0  B1004   
..     ...     ...      ...    ...   ...        ...   ...         ...    ...   
89    None    None     None   None  None       None     5          12  B1060   
90    None    None     None   None  None       None     5          13  B1058   
91    None    None     None   None  None       None     5          12  B1051   
92    None    None     None   None  None       None     5          12  B1060   
93    None    None     None   None  None       None     5           8  B1062   

   Longitude Latitude  
4   -80.5774  28.5619  
5   -80.5774  28.5619  
6   -80.5774  28.5619  
7   -120.611  34.6321  
8   -80.5774  28.5619  
..       ...      ...  
89   -80.604  28.6081  
90   -80.604  28.6081  
91   -80.604  28.6081  
92  -80.5774  28.5619  
93  -80.5774  28.5619  

[90 rows x 17 columns]

## Data Wrangling


We can see below that some of the rows are missing values in our dataset.


In [26]:
#data_falcon9.isnull('LandingPad')
data_falcon9.isnull().sum()

FlightNumber       0
Date               0
BoosterVersion     0
PayloadMass        5
Orbit              0
LaunchSite         0
Outcome           90
Flights           90
GridFins          90
Reused            90
Legs              90
LandingPad        90
Block              0
ReusedCount        0
Serial             0
Longitude          0
Latitude           0
dtype: int64

Before we can continue we must deal with these missing values. The <code>LandingPad</code> column will retain None values to represent when landing pads were not used.


### Task 3: Dealing with Missing Values


Calculate below the mean for the <code>PayloadMass</code> using the <code>.mean()</code>. Then use the mean and the <code>.replace()</code> function to replace `np.nan` values in the data with the mean you calculated.


In [27]:
# Calculate the mean value of PayloadMass column
payloadmassmean = data_falcon9["PayloadMass"].astype("float").mean(axis=0)
print(payloadmassmean)
# Replace the np.nan values with its mean value

6123.547647058824


You should see the number of missing values of the <code>PayLoadMass</code> change to zero.


Now we should have no missing values in our dataset except for in <code>LandingPad</code>.


We can now export it to a <b>CSV</b> for the next section,but to make the answers consistent, in the next lab we will provide data in a pre-selected date range. 


In [28]:
data_falcon9.to_csv('dataset_part_1.csv', index=False)

## Authors


<a href="https://www.linkedin.com/in/joseph-s-50398b136/">Joseph Santarcangelo</a> has a PhD in Electrical Engineering, his research focused on using machine learning, signal processing, and computer vision to determine how videos impact human cognition. Joseph has been working for IBM since he completed his PhD. 


## Change Log


|Date (YYYY-MM-DD)|Version|Changed By|Change Description|
|-|-|-|-|
|2020-09-20|1.1|Joseph|get result each time you run|
|2020-09-20|1.1|Azim |Created Part 1 Lab using SpaceX API|
|2020-09-20|1.0|Joseph |Modified Multiple Areas|


Copyright © 2021 IBM Corporation. All rights reserved.
